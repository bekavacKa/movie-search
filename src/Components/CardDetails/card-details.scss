@import '../../Assets/scss/mixins.scss';
@import '../../Assets/scss/variables.scss';

.card-details{
    @include my-flex(center, row, normal);
    background-position: right center;
    background-repeat: no-repeat;
    background-size: cover;
    padding: 100px;
    .card-details-image{
        flex: 1;
        @include my-flex(center, column, center);
        img{
            border-radius: 5px;
            box-shadow: 0 0 7px 1px $firstFontColor;
            width: 70%;
        }
        .image-homepage{
            width: 70%;
            @include my-flex(center, column, center);
            transition: $transition;
            cursor: pointer;
            img{
                border-radius: 5px;
                box-shadow: 0 0 7px 1px $firstFontColor;
                width: 100%;
            }
            &:hover{
                box-shadow: 0 0 7px 2px $firstFontColor;
                transform: scale(1.01);
            }
        }
    }
    .card-details-info{
        flex: 2;
        @include my-flex(flex-start, column, normal);
        .card-details-title{
            @include my-flex(flex-start, row, center);
            text-transform: capitalize;
            margin-bottom: 30px;
            letter-spacing: .1rem;
            h2{
                font-size: 2rem;
            }
            .title-icon{
                @include my-flex(flex-start, row, center);
                margin: 0 20px;
                .imdb{
                    font-size: 2rem;
                    color: yellow;
                }
                &:hover{
                    transform: scale(1.05);
                }
            }
        }
        .card-details-overview{
            font-size: 1.2rem;
            letter-spacing: .1rem;
            line-height: 1.5rem;
        }
        .card-details-video{
            @include my-flex(flex-start, row, center);
            margin-bottom: 15px;
            font-size: 1rem;
            text-transform: uppercase;
            .video-icon{
                font-size: 1.2rem;
                margin-right: 20px;
                transition: $transition;
                &:hover{
                    transform: scale(1.05);
                }
            }
        }
    }
}

@media only screen and (max-width: 319px) {
    .card-details{
        @include my-flex(center, column, center);
        background-position: center;
        padding: 30px 10px;
        .card-details-image{
            flex: 1;
            img{
                width: 100%;
            }
            .image-homepage{
                width: 100%;
                @include my-flex(center, column, center);
                img{
                    border-radius: 5px;
                    box-shadow: 0 0 7px 1px $firstFontColor;
                    width: 100%;
                }
            }
        }
        .card-details-info{
            flex: 1;
            @include my-flex(center, column, center);
            text-align: center;
            .card-details-title{
                margin: 30px 0;
                @include my-flex(center, column, center);
                .title-icon{
                    margin-top: 10px;
                }
                h2{
                    font-size: 2rem;
                }
            }
            .card-details-overview{
                font-size: 1.1rem;
                line-height: 1.4rem;
            }
        }
    }
}

@media only screen and (min-width: 320px) and (max-width: 768px){
    .card-details{
        @include my-flex(center, column, center);
        background-position: center;
        padding: 30px 10px;
        .card-details-image{
            flex: 1;
            img{
                width: 100%;
            }
            .image-homepage{
                width: 100%;
                @include my-flex(center, column, center);
                img{
                    border-radius: 5px;
                    box-shadow: 0 0 7px 1px $firstFontColor;
                    width: 100%;
                }
            }
        }
        .card-details-info{
            flex: 1;
            @include my-flex(center, column, center);
            text-align: center;
            .card-details-title{
                margin: 30px 0;
                @include my-flex(center, column, center);
                .title-icon{
                    margin-top: 10px;
                }
                h2{
                    font-size: 2rem;
                }
            }
            .card-details-overview{
                font-size: 1.1rem;
                line-height: 1.4rem;
            }
        }
    }
}

@media only screen and (min-width: 769px) and (max-width: 1024px){
    .card-details{
        background-position: center;
        padding: 50px;
        .card-details-image{
            img{
                width: 80%;
            }
            .image-homepage{
                width: 80%;
                @include my-flex(center, column, center);
                img{
                    border-radius: 5px;
                    box-shadow: 0 0 7px 1px $firstFontColor;
                    width: 100%;
                }
            }
        }
    }
}