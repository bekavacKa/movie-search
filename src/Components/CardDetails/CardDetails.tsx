import React, { FC, useEffect, useState } from "react";
import { FaFilm, FaImdb } from "react-icons/fa";
import { useSelector } from "react-redux";
import { useDispatch } from "react-redux";
import { useParams } from "react-router";
import { setLoader } from "../../Redux/loaderSlice";
import MovieTvService from "../../Services/MovieTvService";

import './card-details.scss';

interface IDetails {
  adult: boolean;
  backdrop_path: string;
  first_air_date: string;
  homepage: string;
  id: number;
  in_production: boolean;
  last_air_date: string;
  name: string;
  next_episode_to_air: null;
  number_of_episodes: number;
  number_of_seasons: number;
  original_language: string;
  original_name: string;
  overview: string;
  popularity: number;
  poster_path: string;
  status: string;
  tagline: string;
  type: string;
  vote_average: number;
  vote_count: number;

  budget: number;
  imdb_id: string;
  original_title: string;
  release_date: string;
  revenue: number;
  runtime: number;
  title: string;
  video: boolean;
}

const imagePath: string = `https://www.themoviedb.org/t/p/w220_and_h330_face/`;

const CardDetails: FC = () => {
  const params = useParams();
  const id = Number(params.id);
  const dispatch = useDispatch();

  const { tvShowsButton } = useSelector((state: any) => state.buttonsStore);
  const { movieButton } = useSelector((state: any) => state.buttonsStore);

  const [details, setDetails] =  useState<IDetails | null>(null);
  const [errMsg, setErrMsg] = useState<string>("")

  useEffect(() => {
    getDetails();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  const getDetails = () => {
    // console.log(id);
    if (tvShowsButton){
      return tvShowsDetails();
    }
    if(movieButton){
      return movieDetails();
    }
  };

  const tvShowsDetails = (): void => {
    dispatch(setLoader(true));
    MovieTvService.getTvShowDetails(id)
      .then((res) => {
        // console.log(res.data);
        setDetails(res.data);
      })
      .catch((err) => {
        console.log(err);
        err.message && setErrMsg(err.message);
      })
      .finally(() => dispatch(setLoader(false)));
  }

  const movieDetails = (): void => {
    dispatch(setLoader(true));
    MovieTvService.getMovieDetails(id)
                  .then((res) => {
                    // console.log(res.data);
                    setDetails(res.data);
                  })
                  .catch((err) => {
                    console.log(err);
                    err.message && setErrMsg(err.message);
                  })
                  .finally(() => dispatch(setLoader(false)));
  }

  return (
    <div className="card-details" style={{backgroundImage: `linear-gradient( rgba(0, 0, 0, 0.9), rgba(0, 0, 0, 0.9)), url(${imagePath + details?.backdrop_path})`}}>


      {
        details?.homepage ?
        <div className="card-details-image">
        {
          details?.poster_path ?
          <a className="image-homepage" href={details?.homepage} target="_blank" rel="noopener noreferrer">
            <img src={`${imagePath + details?.poster_path}`} alt={details?.name} />
          </a>
          :
          <a className="image-homepage" href={details?.homepage} target="_blank" rel="noopener noreferrer">
            <img src={`${imagePath + details?.backdrop_path}`} alt={details?.name} />
          </a>
        }
        </div>
        :
        <div className="card-details-image">
          {
            details?.poster_path ?
            <img src={`${imagePath + details?.poster_path}`} alt={details?.name} />
            :
            <img src={`${imagePath + details?.backdrop_path}`} alt={details?.name} />
          }
        </div>
      }

      <div className="card-details-info">
        <div className="card-details-title">
          <h2>{details?.name || details?.original_title}</h2>
          {
            details?.imdb_id &&
            <a className="title-icon" href={`https://www.imdb.com/title/${details?.imdb_id}/?ref_=vp_vi_tt`} target="_blank" rel="noopener noreferrer">
              <FaImdb className="imdb" />
            </a>
          }
        </div>

        {
          details?.imdb_id &&
          <a className="card-details-video" href={`https://www.imdb.com/video/vi1348706585/?playlistId=${details?.imdb_id}&ref_=tt_ov_vi`} target="_blank" rel="noopener noreferrer">
            <FaFilm className="video-icon" />
            <p>thriller</p>
          </a>
        }

        <div className="card-details-overview">
          <p>{details?.overview}</p>
        </div>


      </div>


      {
        errMsg &&
        <div className="card-details-err">
          <p>Sorry, we did not find any details for the selected TV show / movie!</p>
        </div>
      }
    </div>

  )
}

export default CardDetails;
