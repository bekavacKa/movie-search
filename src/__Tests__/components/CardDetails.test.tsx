import renderer from 'react-test-renderer';
import {BrowserRouter as Router} from 'react-router-dom';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import CardDetails from '../../Components/CardDetails/CardDetails';

test('Snapshot test', () => {
    const initialState = { 
        searchTermStore: '',
        buttonsStore:{
            tvShowsButton: true,
            movieButton: false
        },
        storeSearchTerm: ""
    };
    const mockStore = configureStore();
    let store;

    store = mockStore(initialState);
  const tree = renderer.create(
    <Provider store={store}>
        <Router>
            <CardDetails />
        </Router>
    </Provider>
    ).toJSON();
  expect(tree).toMatchSnapshot();
});
