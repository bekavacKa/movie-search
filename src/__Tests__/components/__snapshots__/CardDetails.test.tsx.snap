// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot test 1`] = `
<div
  className="card-details"
  style={
    {
      "backgroundImage": "linear-gradient( rgba(0, 0, 0, 0.9), rgba(0, 0, 0, 0.9)), url(https://www.themoviedb.org/t/p/w220_and_h330_face/undefined)",
    }
  }
>
  <div
    className="card-details-image"
  >
    <img
      src="https://www.themoviedb.org/t/p/w220_and_h330_face/undefined"
    />
  </div>
  <div
    className="card-details-info"
  >
    <div
      className="card-details-title"
    >
      <h2 />
    </div>
    <div
      className="card-details-overview"
    >
      <p />
    </div>
  </div>
</div>
`;
